### ISSUE TITLE
question about addIterativeLSI

### ISSUE THREAD

**Issue by sinihak** [GreenleafLab/ArchR#540]
Hi!

I'm having some issues with the addInterativeLSI function and especially the additional 
parameters defined in clusterParams. Firstly, I am having trouble finding proper values for the parameters. 
When trying to change the resolution or sampleCells, I often get this error:

![image](https://user-images.githubusercontent.com/57436442/108479499-7c418500-729e-11eb-8479-66cb657375b4.png)

Secondly, when I find the values with which the function runs through, it assigns all the cells in my
samples in one cluster. In Signac, findClusters() was able to find several clusters in the same samples as
well as cisTopic's clustering tool. If I leave the additional arguments undefined, the function works, but still, 
there's only one cluster.

I am unsure, whether I am using proper values for the arguments (especially resolution, sampleCell 
and varFeatures. Do you have any suggestions, how to choose the values to best describe my data?

R version 3.6.1 (2019-07-05)
Platform: x86_64-conda_cos6-linux-gnu (64-bit)
Running under: CentOS Linux 7 (Core)

Matrix products: default
BLAS/LAPACK: /research/users/sinihak/miniconda3/envs/sc-tutorial/lib/R/lib/libRblas.so

locale:
 [1] LC_CTYPE=en_US.UTF-8          LC_NUMERIC=fi_FI.UTF-8       
 [3] LC_TIME=fi_FI.UTF-8           LC_COLLATE=en_US.UTF-8       
 [5] LC_MONETARY=fi_FI.UTF-8       LC_MESSAGES=en_US.UTF-8      
 [7] LC_PAPER=fi_FI.UTF-8          LC_NAME=en_US.UTF-8          
 [9] LC_ADDRESS=en_US.UTF-8        LC_TELEPHONE=en_US.UTF-8     
[11] LC_MEASUREMENT=fi_FI.UTF-8    LC_IDENTIFICATION=en_US.UTF-8

attached base packages:
 [1] grid      parallel  stats4    stats     graphics  grDevices utils    
 [8] datasets  methods   base     

other attached packages:
 [1] gridExtra_2.3               nabor_0.5.0                
 [3] Seurat_3.2.2                uwot_0.1.8                 
 [5] Cairo_1.5-10                hdf5r_1.3.3                
 [7] ArchR_1.0.0                 magrittr_2.0.1             
 [9] rhdf5_2.30.1                Matrix_1.2-18              
[11] data.table_1.13.6           SummarizedExperiment_1.16.1
[13] DelayedArray_0.12.3         BiocParallel_1.20.1        
[15] matrixStats_0.57.0          Biobase_2.46.0             
[17] GenomicRanges_1.38.0        GenomeInfoDb_1.22.1        
[19] IRanges_2.20.2              S4Vectors_0.24.4           
[21] BiocGenerics_0.32.0         ggplot2_3.3.3              

loaded via a namespace (and not attached):
  [1] Rtsne_0.15             colorspace_2.0-0       deldir_0.2-3          
  [4] ellipsis_0.3.1         ggridges_0.5.2         IRdisplay_0.7.0       
  [7] XVector_0.26.0         base64enc_0.1-3        spatstat.data_1.5-2   
 [10] farver_2.0.3           leiden_0.3.5           listenv_0.8.0         
 [13] ggrepel_0.8.2          bit64_4.0.5            RSpectra_0.16-0       
 [16] codetools_0.2-18       splines_3.6.1          polyclip_1.10-0       
 [19] IRkernel_1.1.1         jsonlite_1.7.2         ica_1.0-2             
 [22] cluster_2.1.0          png_0.1-7              shiny_1.5.0           
 [25] sctransform_0.3.1      compiler_3.6.1         httr_1.4.2            
 [28] fastmap_1.0.1          lazyeval_0.2.2         later_1.1.0.1         
 [31] htmltools_0.5.0        tools_3.6.1            rsvd_1.0.3            
 [34] igraph_1.2.6           gtable_0.3.0           glue_1.4.2            
 [37] GenomeInfoDbData_1.2.2 RANN_2.6.1             reshape2_1.4.4        
 [40] dplyr_1.0.2            spatstat_1.64-1        Rcpp_1.0.5            
 [43] vctrs_0.3.6            nlme_3.1-148           lmtest_0.9-38         
 [46] stringr_1.4.0          globals_0.13.1         mime_0.9              
 [49] miniUI_0.1.1.1         lifecycle_0.2.0        irlba_2.3.3           
 [52] gtools_3.8.2           goftest_1.2-2          future_1.20.1         
 [55] zlibbioc_1.32.0        MASS_7.3-53            zoo_1.8-8             
 [58] scales_1.1.1           spatstat.utils_1.17-0  promises_1.1.1        
 [61] RColorBrewer_1.1-2     reticulate_1.18        pbapply_1.4-3         
 [64] rpart_4.1-15           stringi_1.5.3          repr_1.1.0            
 [67] rlang_0.4.10           pkgconfig_2.0.3        bitops_1.0-6          
 [70] evaluate_0.14          lattice_0.20-41        tensor_1.5            
 [73] ROCR_1.0-11            purrr_0.3.4            Rhdf5lib_1.8.0        
 [76] labeling_0.4.2         patchwork_1.1.0        htmlwidgets_1.5.3     
 [79] cowplot_1.1.0          bit_4.0.4              tidyselect_1.1.0      
 [82] parallelly_1.21.0      RcppAnnoy_0.0.18       plyr_1.8.6            
 [85] R6_2.5.0               generics_0.1.0         pbdZMQ_0.3-3.1        
 [88] mgcv_1.8-33            pillar_1.4.7           withr_2.3.0           
 [91] fitdistrplus_1.1-1     abind_1.4-5            survival_3.2-7        
 [94] RCurl_1.98-1.2         tibble_3.0.4           future.apply_1.6.0    
 [97] crayon_1.3.4           uuid_0.1-4             KernSmooth_2.23-18    
[100] plotly_4.9.3           digest_0.6.27          xtable_1.8-4          
[103] tidyr_1.1.2            httpuv_1.5.4           munsell_0.5.0         
[106] viridisLite_0.3.0

**Comment by rcorces** [comment]
In the future, please use the provided issue templates. There is a reason that they exist. This is also not a bug report.

To clarify, `ArchR::addIterativeLSI()` calls `ArchR::addClusters()` which calls `Seurat::FindClusters()`. So it is calling the same clustering function that you refer to with Signac. The `clusterParams` are therefore defined by `Seurat::FindClusters()` so you can see what options are available here:
https://www.rdocumentation.org/packages/Seurat/versions/1.4.0/topics/FindClusters

However, your description raises another point of clarification. Clusters called in `addIterativeLSI()` are only used for refining the dimensionality reduction. These clusters are not stored in your dataset. That is done using a separate call to `addClusters()` after performing dimensionality reduction. This is explained in detail in the manual:
https://www.archrproject.com/bookdown/clustering-using-seurats-findclusters-function.html

Please close if this answers your question.

**Comment by sinihak** [comment]
Hi, this didn't really answer my question. As I said, the initial clustering of this same dataset works fine with Signac and cisTopic, so I suppose this tool is just not proper for my data, because Archr is not able to find but one cluster with the addClusters() function. So, to clarify, even after numerous tries with different argument values to the function my persistent issue is that Archr only manages to find one (1) cluster whereas Signac and cisTopic manages with ease.

**Comment by rcorces** [comment]
Perhaps you could post images of your UMAPs from ArchR and Signac so we can get a better sense of what you mean.

**Comment by sinihak** [comment]
Hi, the first is from Signac and latter from Archr:

![image](https://user-images.githubusercontent.com/57436442/108629923-e1e26c80-746a-11eb-98fe-7241b22dad26.png)
 
![image](https://user-images.githubusercontent.com/57436442/108629967-17875580-746b-11eb-9116-2873a2b60b80.png)
 
The clusters did not separate very well in Signac either.

**Comment by rcorces** [comment]
Maybe @jgranja24 can weigh in. 
I'm sure tweaking some of the parameters to `addIterativeLSI()` might help. But its also worth noting that more clusters may not be good if there is no biological relevance to the clusters. So if the "clusters" identified in the Signac dim. reduction are artifactual, then that would be a bad thing. 

Just glancing at the code you pasted in the original post, 10 iterations is a lot, this is normally set to 2 or 3. I imagine you tried those though and were just experimenting with other parameter options.

I normally change the following parameters:
```
iterations= 2 or 3
cluster resolution = 0.05, 0.1, 0.2, 0.3, 0.4, 0.5, 1 2
varFeatures= 10000, 20000, 30000
```

**Comment by jgranja24** [comment]
To add to that you can also increase varFeatures eg 50000 and 100000. Do you have the params you used for Signac as well?

**Comment by sinihak** [comment]
Yes, I have tried also lower number of iterations. I have tried the same parameters that worked in Signac (e.g. resolution, dimensions) but they have resulted in error. However, I noticed that the varFeatures i've been quite low considering my samples and topFeatures (with min.cutoff = 'q0' i.e all features) so that is something I will need to change and also try your suggestions concerning other arguments. Thanks for helping me! :)
